### PR TITLE
Rule S4426: support more cryptographic algorithms

### LINKED ISSUE #3097: Rule S4426 improvements: support of more cryptographic algorithms
Implements [RSPEC-4426](https://jira.sonarsource.com/browse/RSPEC-4426)

Code examples:
https://github.com/SonarSource/security-expected-issues/tree/master/dotnet/wip/basic-rules/RSPEC_4426

### PR DESCRIPTION
**Pull request by christophe-zurn-sonarsource**
Fixes #3097

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs [hunk 1]
```diff
@@ -34,42 +35,243 @@ namespace SonarAnalyzer.Rules.CSharp
...
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, syntaxElement.GetLocation(), MinimalEllipticCurveKeyLength, "EC", ""));
+            }
+        }
+
+        private static void CheckSystemSecurityCryptographyAlgorithms(ITypeSymbol containingType, ObjectCreationExpressionSyntax objectCreation, SyntaxNodeAnalysisContext c)
+        {
+            var firstParam = objectCreation.ArgumentList.Get(0);
+
+            // DSACryptoServiceProvider is always noncompliant as it has a max key size of 1024
+            // RSACryptoServiceProvider default constructor is noncompliant as it has default key size of 1024
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: This will need to be updated for `(parameters)` overload.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs [hunk 2]
```diff
@@ -34,42 +35,243 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class CryptographicKeyShouldNotBeTooShort : SonarDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S4426";
-        private const string MessageFormat = "Use a key length of at least '{0}' bits.";
+        private const string MessageFormat = "Use a key length of at least {0} bits for {1} cipher algorithm.{2}";
+        private const string AdditionalInfo = " This assignment does not actually update the underlying key size.";
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: This constant name seems to generic and it contains clear (single purpose) message. I suggest to rename it to something like `UselessAssignmentInfo`

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs [hunk 3]
```diff
@@ -34,42 +35,243 @@ namespace SonarAnalyzer.Rules.CSharp
...
+                {
+                    var assignment = (AssignmentExpressionSyntax)c.Node;
+                    var propertyName = GetPropertyName(assignment.Left);
+                    if (propertyName == "KeySize"
+                        && assignment.Left is MemberAccessExpressionSyntax memberAccess
+                        && memberAccess.Expression != null
+                        && c.SemanticModel.GetTypeInfo(memberAccess.Expression).Type is ITypeSymbol containingType)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectCreation.GetLocation(), MinimalKeyLengthForRsa));
+                        // Using the KeySize setter on DSACryptoServiceProvider/RSACryptoServiceProvider does not actually change the underlying key size
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: This is serious charge. It would be nice to have link to MS docs on second line.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs [hunk 4]
```diff
@@ -11,10 +19,10 @@ public class Program
         private static readonly int validKeySize = 2048;
         private static readonly int invalidKeySize = 1024;
 
-        public void Method()
+        public void ConstArgumentResolution()
         {
             const int localValidSize = 2048;
-            new RSACryptoServiceProvider();
+            new RSACryptoServiceProvider(); // Noncompliant {{Use a key length of at least 2048 bits for RSA cipher algorithm.}}
             new RSACryptoServiceProvider(new CspParameters());
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: Good catch 👍 
This was just omitted

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs [hunk 5]
```diff
@@ -24,11 +32,468 @@ public void Method()
...
+        {
+            var ec1 = new ECDiffieHellmanCng(); // Compliant - default EC key size is 521
+            var ec2 = new ECDsaCng(); // Compliant - default EC key size is 521
+            var ec3 = new ECDsaOpenSsl(); // Compliant - default EC key size is 521
+
+            // Valid key sizes are 256, 384, and 521 bits.
+            var ec4 = new ECDiffieHellmanCng(256);
+            var ec5 = new ECDsaCng(384);
+            var ec6 = new ECDsaOpenSsl(521);
+            var ec7 = new ECDsaOpenSsl((IntPtr)128);
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: No this is not a FN, it's actually to ensure the rule does not raise for this. The param type is `IntPtr` (although it does not make sense to use `IntPtr` on a constant). The `IntPtr` should actually point to the address of the key

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs [hunk 6]
```diff
@@ -24,11 +32,468 @@ public void Method()
...
+            var ec1 = new ECDiffieHellmanCng(ECCurve.NamedCurves.brainpoolP192r1); // Noncompliant {{Use a key length of at least 224 bits for EC cipher algorithm.}}
+            var ec2 = new ECDsaCng(ECCurve.NamedCurves.brainpoolP160t1); // Noncompliant {{Use a key length of at least 224 bits for EC cipher algorithm.}}
+            var ec3 = new ECDsaOpenSsl(ECCurve.NamedCurves.brainpoolP192t1); // Noncompliant
+        }
+
+        public void CompliantKeySizeSet()
+        {
+            var ec1 = new ECDiffieHellmanCng();
+            ec1.KeySize = 512;
+            ec1.KeySize = 128; // OK - because this is not a valid key size for this object
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: TBH, this is for implementation simplicity: it's easier for DSACryptoServiceProvider to raise an issue all the time (since it's always noncompliant), and here it's easier to never raise an issue (sinced it's can only be compliant). And since these are edge cases (invalid use of these constructors), I would not expect it to happen on real code anyways

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs [hunk 7]
```diff
@@ -24,11 +32,468 @@ public void Method()
...
+            var dsa2 = new DSAOpenSsl(2048);
+        }
+
+        public void NonCompliantKeySizeConstructors(CspParameters parameters)
+        {
+            var dsa1 = new DSACng(512); // Noncompliant {{Use a key length of at least 2048 bits for DSA cipher algorithm.}}
+            var dsa2 = new DSACryptoServiceProvider(512); // Noncompliant {{Use a key length of at least 2048 bits for DSA cipher algorithm.}}
+            var dsa3 = new DSACryptoServiceProvider(1024, parameters); // Noncompliant
+            var dsa4 = new DSACryptoServiceProvider(3072); // Noncompliant - this is not valid
+            var dsa5 = new DSACryptoServiceProvider(2048, parameters); // Noncompliant - this is not valid
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: I agree. But I think it is overkill to add an addtional message for a case that we should actually never encounter (since this is not a valid case). So, for simplicity purpose, I'd suggest to keep it as is

File: sonaranalyzer-dotnet/rspec/cs/S4426_c#.html [hunk 8]
```diff
@@ -13,7 +28,16 @@ <h2>Noncompliant Code Example</h2>
...
         static void Main()
         {
-            RSACryptoServiceProvider RSA = new RSACryptoServiceProvider(1024); // Noncompliant
+            var dsa1 = new DSACryptoServiceProvider(); // Noncompliant - default key size is 1024
+            dsa1.KeySize = 2048; // Noncompliant - the setter does not actually update the underlying key size for the DSACryptoServiceProvider class
+
+            var dsa2 = new DSACryptoServiceProvider(2048); // Noncompliant - cannot create DSACryptoServiceProvider with a key size bigger than 1024
+
+            var rsa1 = new RSACryptoServiceProvider(); // Noncompliant - default key size is 1024
+            rsa1.KeySize = 2048; // Noncompliant - the setter does not actually update the underlying key size for the RSACryptoServiceProvider class
```
[the remark this answers is not in the document]
Comment (christophe-zurn-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S4426_c#.html]: Ok with me to remove the `actually` part.
After discussion with @eric-therond-sonarsource , we decided to mention it in the RSPEC examples. I actually added the message part after loooking at inital results, as I felt it was important to report something more in that case.

I let you discuss with @eric-therond-sonarsource tomorrow to see whether it is worth updating RSPEC to mention that case in the description as well as to add doc links

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: Ok, considering that the message asks to replace the entire class.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CryptographicKeyShouldNotBeTooShort.cs]: Then we need to update the comment on line 174

File: sonaranalyzer-dotnet/rspec/cs/S4426_c#.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/rspec/cs/S4426_c#.html]: Eric agreed to mention this case as paragraph in RSPEC.
